### ISSUE TITLE
How to update plugins inside logstash container

### ISSUE THREAD

**Issue by DamionWaltermeyer** [deviantony/docker-elk#57]
Howdy.
I'm trying to use this docker netflow v10/ipfix.    The logstash-codec-netflow present is using 2.0.5. (gems/logstash-codec-netflow-2.0.5) 

I tried adding the line to the dockerfile, but since it's already present as an old version, It doesn't update to 3.1.0 (current).     

I'm getting an error when I try to use the logstash update. 
```
root@be651f0dbb6f:/opt/logstash/bin# logstash-plugin update logstash-codec-netflow        
You are updating logstash-codec-netflow to a new version 3.1.0, which may not be compatible with 2.0.5. are you sure you want to proceed (Y/N)?
Y
Updating logstash-codec-netflow
Error Bundler::InstallError, retrying 1/10
An error occurred while installing ffi (1.9.13), and Bundler cannot continue.
Make sure that `gem install ffi -v '1.9.13'` succeeds before bundling.
WARNING: SSLSocket#session= is not supported
```
At first, I went through and found gem wasn't installed, so then I had to install ruby. I ended up with 
```
root@be651f0dbb6f:/opt# whereis ruby
ruby: /usr/bin/ruby /usr/bin/ruby2.1 /usr/lib/ruby /usr/share/man/man1/ruby.1.gz
root@be651f0dbb6f:/opt# whereis gem 
gem: /usr/bin/gem2.1 /usr/bin/gem /usr/share/man/man1/gem.1.gz
root@be651f0dbb6f:/opt/logstash/bin# gem install ffi -v '1.9.13'                       
Fetching: ffi-1.9.13.gem (100%)
Building native extensions.  This could take a while...
ERROR:  Error installing ffi:
	ERROR: Failed to build gem native extension.

    current directory: /var/lib/gems/2.1.0/gems/ffi-1.9.13/ext/ffi_c
/usr/bin/ruby2.1 -r ./siteconf20160707-912-ebyoz5.rb extconf.rb
mkmf.rb can't find header files for ruby at /usr/lib/ruby/include/ruby.h

extconf failed, exit code 1

Gem files will remain installed in /var/lib/gems/2.1.0/gems/ffi-1.9.13 for inspection.
Results logged to /var/lib/gems/2.1.0/extensions/x86_64-linux/2.1.0/ffi-1.9.13/gem_make.out
root@be651f0dbb6f:/opt/logstash/bin# logstash-plugin update logstash-codec-netflow     
You are updating logstash-codec-netflow to a new version 3.1.0, which may not be compatible with 2.0.5. are you sure you want to proceed (Y/N)?
Y
Updating logstash-codec-netflow
Error Bundler::InstallError, retrying 1/10
An error occurred while installing ffi (1.9.13), and Bundler cannot continue.
Make sure that `gem install ffi -v '1.9.13'` succeeds before bundling.
WARNING: SSLSocket#session= is not supported
Error Bundler::InstallError, retrying 2/10

```

Use case for question: 

Since 2.0.5 doesn't support v9 and v10 as well as 3.1.0 now does, I'm getting this pretty exclusively for my logstash log.
```
{:timestamp=>"2016-07-07T19:21:51.463000+0000", :message=>"No matching template for flow id 259", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:51.490000+0000", :message=>"Unsupported Netflow version v10", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:51.909000+0000", :message=>"No matching template for flow id 259", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:51.989000+0000", :message=>"Unsupported Netflow version v10", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:52.457000+0000", :message=>"No matching template for flow id 258", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:52.957000+0000", :message=>"No matching template for flow id 258", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:53.353000+0000", :message=>"No matching template for flow id 258", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:54.054000+0000", :message=>"No matching template for flow id 258", :level=>:warn}
{:timestamp=>"2016-07-07T19:21:54.690000+0000", :message=>"Unsupported Netflow version v10", :level=>:warn}
```

FYI: Rubyn00b.  I broke the container many times and had to rebuild trying to do this. 

Thanks for any assistance.

**Comment by deviantony** [comment]
Hey there ! Sorry never played with netflow, however I made a quick search about the `logstash-codec-netflow` and found this:

> 3.1.0 is only supported for LS 5.x release. 

in this issue: https://github.com/logstash-plugins/logstash-codec-netflow/issues/38

You may want to use the `x-pack` branch and try it against the latest alpha version: https://github.com/deviantony/docker-elk/tree/x-pack

**Comment by deviantony** [comment]
@DamionWaltermeyer any update?

**Comment by DamionWaltermeyer** [comment]
Hi @deviantony, Sorry for the delay in getting back to you. I ended up being out sick for a bit. I'm going to go through it and give it a shot today. Thank you for the info.

**Comment by DamionWaltermeyer** [comment]
Fun stuff.  With the new x-pack branch  there is an update that breaks the netflow config.
https://github.com/logstash-plugins/logstash-codec-netflow/issues/41

So I ended up clearing out the netflow codec from the logstash.conf,  bringing it up with docker-compose, and then updating the netflow codec to fix the netflow bug, then putting the good config back and restarting the container.   If I don't pull the config, the logstash container won't stay up long enough for me to run the update command. 

Now it's having a couple issues it wasn't having before.
```
{:timestamp=>"2016-07-15T16:38:27.894000+0000", :message=>"Encountered an unexpected error submitting a bulk request! Will retry.", :error_message=>"undefined method `code' for #<LogStash::Outputs::ElasticSearch::HttpClient::Pool::BadResponseCodeError:0x72c6c795>", :class=>"NoMethodError", :backtrace=>["/usr/share/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-4.1.1-java/lib/logstash/outputs/elasticsearch/common.rb:217:in `safe_bulk'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-4.1.1-java/lib/logstash/outputs/elasticsearch/common.rb:105:in `submit'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-4.1.1-java/lib/logstash/outputs/elasticsearch/common.rb:72:in `retrying_submit'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-4.1.1-java/lib/logstash/outputs/elasticsearch/common.rb:23:in `multi_receive'", "org/jruby/RubyArray.java:1653:in `each_slice'", "/usr/share/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-4.1.1-java/lib/logstash/outputs/elasticsearch/common.rb:22:in `multi_receive'", "/usr/share/logstash/logstash-core/lib/logstash/output_delegator.rb:136:in `threadsafe_multi_receive'", "/usr/share/logstash/logstash-core/lib/logstash/output_delegator.rb:122:in `multi_receive'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:321:in `output_batch'", "org/jruby/RubyHash.java:1342:in `each'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:321:in `output_batch'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:249:in `worker_loop'", "/usr/share/logstash/logstash-core/lib/logstash/pipeline.rb:210:in `start_workers'"], :level=>:error}
```

Also, on the status page, which is all I can get to, I'm getting 
```
plugin:elasticsearch	 Authentication Exception
```

Just to give more information if it helps with any trouble shooting you want to do on your end (I'll be mentioning it on the logstash forums too) 
logstash.conf
```
input {
   udp {
     port => 5656
     codec => netflow {
       versions => [5,9,10]
           target => ipfix
     }
     type => ipfix
   }
 }


output {
  elasticsearch {
    hosts => "elasticsearch:9200"
        index => "netflow-%{+YYYY.MM.dd}"
  }
}
```

current docker-compose.yml

```
elasticsearch:
  build: elasticsearch/
  ports:
    - "9200:9200"
    - "9300:9300"
  environment:
    ES_JAVA_OPTS: "-Xms1g -Xmx1g"
logstash:
  image: logstash:5
  command: logstash -f /etc/logstash/conf.d/logstash.conf
  volumes:
    - ./logstash/config:/etc/logstash/conf.d
  ports:
    - "5656:5656"
  links:
    - elasticsearch
kibana:
  build: kibana/
  volumes:
    - ./kibana/config/:/opt/kibana/config/
  ports:
    - "5601:5601"
  links:
    - elasticsearch
```

Thanks again for the help.

**Comment by deviantony** [comment]
Unfortunetaly, I do not have the time at the moment to help you debug further... I'll leave this issue open, please update it if you find a solution.

**Comment by DamionWaltermeyer** [comment]
Took a break for some other issues, but I wanted to update.  So after I moved to logstash x-pack, i can't start with netflow-codec in the logstash config, I have to start with a minimal config, just  input/port,   and then i can get in to the logstash container and then run:
```
logstash-plugin update
```
It will go through and update all the plugins,  after it's done that, I can exit the container and go back to the config file using the netflow-codec.     It seems to be accepting after that, though I'm still debugging it's new connection to elasticsearch issue, which I think was really just a matter of making sure to use the new x-pack elasticsearch container and putting a proper username/password in the logstash config. Either way, it's unrelated to this particular issue I was having and I think that is now fixed.  Thanks for your help.
